### PR TITLE
Using area of influence for buffer and trade area analyses

### LINKED ISSUE #7126: Buffer and trade-area analyses should work in the same form/context, area of influence

### PR DESCRIPTION
**Pull request by xavijam**
screen shot 2016-04-13 at 17 24 54

screen shot 2016-04-13 at 17 25 37

We needed to use the same form for buffer and trade-area analyses. All the logic when the analysis type changes are placed within the area-of-influence model. 

From now on, any analysis attribute that doesn't belong to the default analysis params will be placed within a new 'options' context, so for example:

```js
…
{
  id: 'a0',
  type: 'source',
  params: {
    query: 'SELECT * FROM hello'
  },
  options: {
    table_name: 'hello'
  }
}
…
```

This new `options` thing won't be sent to **CartoDB.js** or to **Maps API**, but it will be saved within the Editor.

CR: @javierarce 
cc @alonsogarciapablo 

@juanignaciosl, we will need to not populate new `options` object to Maps API for Named Maps.

Fixes #7126.

### CODE REVIEW

**Comment by xavijam** [comment]
I'm finishing tests for the new area-of-influence model but in the meantime I'd like to know your opinion ;).

File: lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js [hunk 1]
```diff
@@ -53,33 +53,48 @@ module.exports = cdb.core.Model.extend({
...
       } else {
         memo[name] = this.get(name);
       }
 
       return memo;
     }, {}, this);
 
-    var json = _.omit(this.attributes, paramNames);
-    json.params = rawParams;
+    var json = {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js]: Only these parameters will be sent to CartoDB.js, rest of them will be added within options context, if we want to send them.

File: lib/assets/locale/en.json [hunk 2]
```diff
@@ -179,11 +179,17 @@
         "point-in-polygon": "Point in polygon",
         "polygons_source": "Polygon source",
         "points_source": "Point source",
-        "buffer": "Buffer",
+        "area-of-influence": "Area of influence",
```
Comment (javierarce) [inline lib/assets/locale/en.json]: I know that the linter will warn us if there's a duplicate, but now that we are starting to have a big list of lines, wouldn't it make sense to sort them alphabetically?

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses-form-types/area-of-influence/area-of-influence-form-model.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
+var BaseAnalysisFormModel = require('../base-analysis-form-model.js');
+var _ = require('underscore');
+var distanceOnMeters = function (type) {
+  switch (type) {
+    case 'kilometers': return 1000;
+    case 'milles': return 1609.34;
```
Comment (javierarce) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses-form-types/area-of-influence/area-of-influence-form-model.js]: Shouldn't be 'miles'?

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses-form-types/area-of-influence/area-of-influence-form-model.js]: :sweat_smile:

File: lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses-form-types/area-of-influence/area-of-influence-form-model.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+  initialize: function (attrs, opts) {
+    if (!opts.analysisDefinitionNodeModel) throw new Error('analysisDefinitionNodeModel is required');
+    this._analysisDefinitionNodeModel = opts.analysisDefinitionNodeModel;
+    this._setModel();
+    this._adjustMetrics();
+    this.bind('change', _.debounce(this._onChange.bind(this), 500));
+  },
+
+  _adjustMetrics: function () {
```
Comment (javierarce) [inline lib/assets/javascripts/cartodb3/editor/layers/layer-content-views/analyses-form-types/area-of-influence/area-of-influence-form-model.js]: `_adjustMetrics` is a bit generic, isn't it? Maybe we could rename it to `_convertRadioToMeters` (or something similar) and check the condition (`if (this.get('type') === 'buffer')`) outside of the method?

File: lib/assets/javascripts/cartodb3/data/analysis-definition-model.js [hunk 5]
```diff
@@ -27,7 +27,7 @@ module.exports = cdb.core.Model.extend({
   toJSON: function () {
     return {
       id: this.id,
-      analysis_definition: this._analysisDefinitionNodesCollection.get(this.get('node_id')).toJSON()
+      analysis_definition: this._analysisDefinitionNodesCollection.get(this.get('node_id')).toJSON(true)
```
Comment (javierarce) [inline lib/assets/javascripts/cartodb3/data/analysis-definition-model.js]: `toJSON(true)`?

**Comment by javierarce** [comment]
Apart from the two comments, LGTM 👍

**Comment by xavijam** [comment]
Added more tests, some pictures in the description and applied what you mentioned ;).

File: lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js [hunk 6]
```diff
@@ -53,33 +53,48 @@ module.exports = cdb.core.Model.extend({
...
       parsedParams
     );
   },
 
   /**
    * @override {Backbone.prototype.toJSON} unflatten the internal structure to the expected nested JSON data structure.
+   * @param {Boolean} addOptions - Add or not analysis options to the definition
    */
-  toJSON: function () {
+  toJSON: function (addOptions) {
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js]: Why not have a `toJSONWithOptions` (or perhaps a different name, but something that has more meaning than `.toJSON(true)`):

```
toJSONWithOptions: function () {
  var json = this.toJSON();
  var optionsAttrs = _.omit(this.attributes, paramNames.concat('id', 'type'));
  if (!_.isEmpty(optionsAttrs)) {
    json['options'] = optionsAttrs;
  }
}
```

Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js]: Yeah, I was on doubt about using the same method too, so agree with you, I'll change it :+1:.

Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/analysis-definition-node-model.js]: Or, given that `toJSON` is the default method to serialize stuff that is sent to the backend, we could have the `toJSON()` include all the options, and then remove the `options` when we serialize the `analysis-definition-model.js`.
Just some ideas...

**Comment by juanignaciosl** [comment]
@xavijam :eyes:

Now, an analysis definition will only send `type`, `id` and `params` to maps api.

**Comment by xavijam** [comment]
The change looks good @juanignaciosl :+1:

**Comment by juanignaciosl** [comment]
@gfiorav CR this, please.

**Comment by gfiorav** [comment]
📦 for me @juanignaciosl

**Comment by juanignaciosl** [comment]
Backend closes #7128 then. @xavijam nothing else, if it works for you, merge.

**Comment by xavijam** [comment]
Yay! waiting for frontend tests and then :ship: :it:.

**Pull request merged**